Shape parameters

hammannr opened the pull request:
This adds shape parameter support for the blueice extended model (closes #33). Also, some bugs are fixed.

Some of the things I modified:
- Make the data a dict (instead of a list) where the keys are the names of the likelihood term. For toy data, the generate_values are the last entry of the data but for real data this doesn't make sense. So to avoid ambiguities, I think it is better to clearly relate the data with a term of the likelihood (or the generate_values).
- The ``ll.rate_parameters`` are missing the `_rate_multiplier` suffix so I updated the parameter overriding and filtering accordingly.
- Add shifted ER templates (naively shifted up and down 1 and 2 bins) to illustrate shape parameters.
- Update the minuit wrapper (now compatible with v>= 2.21)
- The `errordef` was so far not correctly assigned, which caused trouble with the correct error estimation. I also made the minuit object a member of the class, so we can do `statistical_model.minuit_object.minos()` after the fit to very efficiently compute the (asymmetric) uncertainties of all fit parameters!
- Fixed some indentation bug in the search for the template folder
- Enabled str-type parameter space definitions as before.
- Some adjustments to legacy code just to do some cross-checks and debugging.

You can try it using the following code:

```python
from alea.blueice_extended_model import BlueiceExtendedModel

config_path = "alea/examples/model_configs/unbinned_wimp_statistical_model.yaml"

statistical_model = BlueiceExtendedModel.from_config(config_path)
statistical_model.data = statistical_model.generate_data()
best_fit_res, best_fit_nll = statistical_model.fit(verbose=True)
# best fit of er_band_shift should be around 0

statistical_model = BlueiceExtendedModel.from_config(config_path)
statistical_model.data = statistical_model.generate_data(er_band_shift=1.2)
best_fit_res, best_fit_nll = statistical_model.fit(verbose=True)
# best fit of er_band_shift should be around 1.2
```

hammannr commented:
I now updated the exmple to look for the config in `alea/model_configs` but I'm still in favor of re-introducing the ``alea/examples`` folder for configs and templates.

hammannr commented on alea/models/__init__.py:
```diff
@@ -3,3 +1,5 @@
+from ..examples import gaussian_model
+from ..examples.gaussian_model import *
```
ah right. This shouldn't be here.

hammannr commented on alea/examples/__init__.py:
```diff
@@ -0,0 +1,2 @@
+from . import gaussian_model
+from .gaussian_model import *
```
Ok, now I don't know what you want...

kdund commented on alea/model.py:
```diff
@@ -230,28 +229,27 @@ class MinuitWrap:
+            MinuitWrap(cost, parameters=self.parameters),
             **defaults)
+        m.errordef = Minuit.LIKELIHOOD
         fixed_params = [] if fixed_parameters is None else fixed_parameters
```
Not quite sure where, but this code that worked in main is not working now: 

```
from alea.examples import gaussian_model
simple_model = gaussian_model.GaussianModel()
simple_model.data = simple_model.generate_data(mu=0, sigma=2)
bf, ret = simple_model.fit(sigma=2, mu=2)
bf, ret
```

hammannr replied:
Did it work now (since you marked it as resolved?)

kdund commented on alea/model.py:
```diff
@@ -230,28 +229,27 @@ class MinuitWrap:
             """
 
-            def __init__(self, f, s_args):
+            def __init__(self, f, parameters: Parameters):
```
I think that there are two selfs, the self of statistical model, and the self of MinuitWrap

kdund commented on alea/model.py:
```diff
@@ -230,28 +229,27 @@ class MinuitWrap:
             the dictionary passed to the Minuit call.
```
Can minuitWrap be defined standalone instead of inside this function?

kdund reviewed (changes requested):
At the moment, the changes makes it so that fits change if all parameters are set: 
```
from alea.examples import gaussian_model
simple_model = gaussian_model.GaussianModel()
simple_model.data = simple_model.generate_data(mu=0, sigma=2)
bf, ret = simple_model.fit(sigma=2, mu=2)
bf, ret
```

Fails with 

```
----> 1 m = Minuit(MinuitWrap(cost, simple_model.parameters),mu=0, sigma=0)

File ~/opt/miniconda3/envs/myenv/lib/python3.8/site-packages/iminuit/minuit.py:625, in Minuit.__init__(***failed resolving arguments***)
    617 self._strategy = MnStrategy(1)
    618 self._fcn = FCN(
    619     fcn,
    620     getattr(fcn, "grad", grad),
    621     array_call,
    622     getattr(fcn, "errordef", 1.0),
    623 )
--> 625 self._init_state = _make_init_state(self._pos2var, start, kwds)
    626 self._values = mutil.ValueView(self)
    627 self._errors = mutil.ErrorView(self)

File ~/opt/miniconda3/envs/myenv/lib/python3.8/site-packages/iminuit/minuit.py:2470, in _make_init_state(pos2var, args, kwds)
   2468     for kw in kwds:
   2469         if kw not in pos2var:
-> 2470             raise RuntimeError(
   2471                 f"{kw} is not one of the parameters [{' '.join(pos2var)}]"
   2472             )
   2473     nargs = len(kwds)
   2475 if len(pos2var) != nargs:

RuntimeError: mu is not one of the parameters []
```


We'll have to short-circuit this possibility somehow

hammannr commented:
Ah @kdund I think this is an iminuit version problem. I developed with an environment, which contained python 3.0 and iminuit 2.22 and then I got a lot of warnings so I updated the implementation. Now however it only works with iminuit >= 2.21, I also updated the requirements.txt accordingly.

kdund commented on alea/utils.py:
```diff
@@ -10,15 +10,18 @@ def get_analysis_space(analysis_space: dict) -> list:
     for element in analysis_space:
         for key, value in element.items():
-            if value.startswith("np."):
+            if isinstance(value, str) and value.startswith("np."):
```
Do you have an example for usage of this?

hammannr replied:
The example config file has the analysis space defined in the following way:
```python
analysis_space:
        - "cs1": 'np.arange(0, 102, 2)'
        - "cs2": 'np.geomspace(100, 100000, 51)'
```
This will be evaluated accordingly and saves a lot of space compared to the implementation in binference. However, I also re-enabled the binference-like definition, which can be useful when dealing with irregular binnings. To test this you could replace the definition of cs1 with the following:
```python
- "cs1": "0. 2.   4.   6.   8.  10.  12.  14.  16.  18.  20.  22.  24.  26.  28.
  30.  32.  34.  36.  38.  40.  42.  44.  46.  48.  50.  52.  54.  56.
  58.  60.  62.  64.  66.  68.  70.  72.  74.  76.  78.  80.  82.  84.
  86.  88.  90.  92.  94.  96.  98. 100."

```
and it should do exactly the same thing.

hammannr replied:
Also, the example I posted in slack features both implementations (irregular binning in R).

dachengx commented on alea/models/blueice_extended_model.py:
```diff
@@ -84,17 +87,18 @@
-            call_args = {k: i for k, i in kwargs.items() if k in ll_pars}
+        # TODO: Make a self.likelihood_temrs dict with the likelihood names as keys and the corresponding likelihood terms as values.
+        for ll_term, parameter_names in zip(self_copy._likelihood.likelihood_list[:-1],
+                                            self_copy._likelihood.likelihood_parameters):  # ancillary likelihood does not contribute
```
In this case and other similar cases, the comment should be put in another line, and the indent should be changed, according to https://legacy.python.org/dev/peps/pep-0008/#indentation. 

```
        # ancillary likelihood does not contribute
        for ll_term, parameter_names in zip(
            self_copy._likelihood.likelihood_list[:-1],
            self_copy._likelihood.likelihood_parameters):
```

hammannr replied:
Now that we're talking indents: I think we should in this case do one more indent otherwise we'll get "continuation line with same indent as next logical line (E125)".. And this makes code terrible to read.

hammannr replied:
Ah ok nice it even makes this suggestion in the ref you sent below 😄

dachengx commented on alea/models/blueice_extended_model.py:
```diff
@@ -84,17 +87,18 @@ def get_expectation_values(self, **kwargs) -> dict:
+        # TODO: Make a self.likelihood_temrs dict with the likelihood names as keys and the corresponding likelihood terms as values.
+        for ll_term, parameter_names in zip(self_copy._likelihood.likelihood_list[:-1],
+                                            self_copy._likelihood.likelihood_parameters):  # ancillary likelihood does not contribute
+            call_args = {k: i for k, i in generate_values.items() if k in parameter_names}  # WARNING: This silently drops parameters it can't handle!
```
Please move the comment to another line.

dachengx commented on alea/models/blueice_extended_model.py:
```diff
@@ -265,10 +276,10 @@ def __init__(self, parameters: Parameters):
-            func=self.ancillary_likelihood_sum,
-            parameter_list=parameter_list,
-            config=self.parameters.nominal_values)
+        super().__init__(func=self.ancillary_likelihood_sum,
```
I somehow particularly prefer the hanging indent, like at the main branch: Hanging indentation is a type-setting style where all the lines in a paragraph are indented except the first line. Because this will reduce the length of lines. 

Reference: https://legacy.python.org/dev/peps/pep-0008/#indentation

hammannr replied:
I don't mind the hanging indent but here it's not needed since the lines are not too long. We can make some pep8 PR at some point but I would like to not go overboard now with style choices in this PR.

dachengx commented on alea/utils.py:
```diff
@@ -15,15 +15,18 @@ def get_analysis_space(analysis_space: dict) -> list:
+            if isinstance(value, str) and value.startswith("np."):
                 eval_element = (key, eval(value))
+            elif isinstance(value, str):
+                eval_element = (key,
```
Here I also prefer the hanging indentation, like at the main branch.

hammannr replied:
Since it's like this in the main I'll agree to this one. It does look more readable!

dachengx commented on requirements.txt:
```diff
@@ -3,6 +3,7 @@ git+https://github.com/JelleAalbers/blueice
 h5py
 iminuit
 git+https://github.com/XENONnT/inference_interface
+iminuit >= 2.21
```
Usually, we only need `iminuit>=2.21`. Reference: https://pip.pypa.io/en/stable/user_guide/#requirements-files

I also prefer `iminuit>=2.21.0` to indicate that `iminuit` has a patch version.

hammannr replied:
You mean instead of writing `iminuit >= 2.21` write `iminuit>=2.21.0`?

dachengx replied:
Yes.

hammannr replied:
ok fine

dachengx commented on alea/template_source.py:
```diff
@@ -125,8 +125,8 @@ def build_histogram(self):
-        h *= self.config.get('histogram_multiplier', 1)
-        logging.debug(f"Multiplying histogram with histogram_multiplier {self.config.get('histogram_multiplier', 1)}. Histogram is now normalised to {h.n}.")
+        h *= self.config.get('histogram_scale_factor', 1)
+        logging.debug(f"Multiplying histogram with histogram_scale_factor {self.config.get('histogram_scale_factor', 1)}. Histogram is now normalised to {h.n}.")
```
This line is too long. 

```
logging.debug(
    f"Multiplying histogram with histogram_scale_factor {self.config.get('histogram_scale_factor', 1)}."
    f"Histogram is now normalised to {h.n}.")
```

or something else.

hammannr replied:
I think this anyway goes (all the debugging) so I wouldn't bother now cleaning up legacy code.

dachengx commented on alea/models/blueice_extended_model.py:
```diff
@@ -153,25 +160,28 @@ def _build_ll_from_config(self, likelihood_config: dict) -> "LogLikelihoodSum":
 
+                # Set shape parameters
+                shape_parameters = [
+                    p for p in source["parameters"] if self.parameters[p].ptype == "shape"]
```
This indent looks good to me. We can not find other ways to break the line. Also, the report of reviewdog might be a false positive, related to https://github.com/wemake-services/wemake-python-styleguide/issues/1416.

hammannr commented:
@dachengx is now reviewdog 2.0 😄 But doesn't harm to be reminded to write readable code. I'll make those changes once the rest is fixed and the tests are running.

dachengx commented on alea/model_configs/unbinned_wimp_statistical_model.yaml:
```diff
@@ -45,24 +45,26 @@ parameter_definition:
 likelihood_config:
   likelihood_weights: [1, 1, 1]
-  template_folder: []
+  template_folder: alea/templates
```
This one is not needed when we read a file from alea internal location. We can even delete this entire line.

dachengx commented on alea/utils.py:
```diff
@@ -81,6 +83,18 @@ def _package_path(sub_directory):
+    this function replace the parameter with asterisk.
+    """
+    asterisked = formatted
+    for found in re.findall("\{(.*?)\}", formatted):
```
This idea is borrowed from https://stackoverflow.com/questions/27881366/regular-expressions-and.

hammannr replied:
Awesome, that's a very clean solution! 😊

hammannr commented:
Thanks a lot, @dachengx for fixing the wildcard issue -- very nice and clean solution! 😊 I now also implemented some of the formatting suggestions. Do you have any other suggestions regarding the code and have you tried running it?

Also, @kdund did changing the iminuit version solve the issue and are you happy with the implementation of the `analysis_space`? 😊

dachengx commented on alea/utils.py:
```diff
@@ -140,6 +141,8 @@ def get_template_folder_list(likelihood_config):
         template_folder_list = [likelihood_config["template_folder"]]
     elif isinstance(likelihood_config["template_folder"], list):
         template_folder_list = likelihood_config["template_folder"]
+    elif likelihood_config["template_folder"] is None:
```
Great that you find this case!

dachengx reviewed (approved):
Thanks @hammannr . Looks good to me. I agree that we put the code-style cleaning in another PR.

kdund reviewed (approved):
The fitting works for me w/ updated iminuit, thanks, @hammannr!
Slicing and stuff we can look at later.
